File: Anul I - Licenta/Semestrul I/Programarea Algoritmilor/Laboratoare/Laboratorul 6 & 7/14. Reactivi/reactivi.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
using namespace std;
typedef pair<int, int> TemperatureRange;

bool sortCond(const TemperatureRange& a, const TemperatureRange& b)
{
    return (a.second < b.second) || (a.second == b.second && a.first < b.first);
}

int main()
{
    const string inputFileName = "reactivi.in";
    const string outputFileName = "reactivi.out";
    ifstream inputFile(inputFileName);

    int refrigeratorsNumber;
    inputFile >> refrigeratorsNumber;

    vector<TemperatureRange> temperatures;

    for (int i = 0; i < refrigeratorsNumber; ++i)
    {
        int x, y;
        inputFile >> x >> y;
        temperatures.emplace_back(x, y);
    }

    inputFile.close();

    sort(temperatures.begin(), temperatures.end(), sortCond);

    int neededRefrigerators = 1;
    int last = 0;

    for (int index = 1; index < refrigeratorsNumber; ++index)
    {
        if (temperatures[index].first > temperatures[last].second)
        {
            last = index;
            ++neededRefrigerators;
        }
        else
        {
            temperatures[last] = make_pair(
                max(temperatures[index].first, temperatures[last].first),
                min(temperatures[index].second, temperatures[last].second)
            );
        }
    }

    ofstream outputFile(outputFileName);
    outputFile << neededRefrigerators << endl;
    outputFile.close();

    return 0;
}

File: Anul I - Licenta/Semestrul I/Programarea Algoritmilor/Laboratoare/Laboratorul 6 & 7/14. Reactivi/reactivi.py
# Pentru editorul de text
# import os
# path = os.path.dirname(__file__)
# inputFileName = path + "/reactivi.in"
# outputFileName = path + "/reactivi.out"

inputFileName = "reactivi.in"
outputFileName = "reactivi.out"

with open(inputFileName, "r") as f:
    refrigeratorsNumber = int(f.readline().strip())
    temperatures = []
    for line in f.readlines():
        x, y = line.split()
        temperatures.append((int(x), int(y)))

temperatures = sorted(temperatures, key=lambda item: (item[1], item[0]))

neededRefrigerators = index = 1
last = 0
while index < refrigeratorsNumber:
    if temperatures[index][0] > temperatures[last][1]:
        last = index
        neededRefrigerators += 1
    else:
        temperatures[last] = (max(temperatures[index][0], temperatures[last][0]), min(temperatures[index][1], temperatures[last][1]))

    index += 1

with open(outputFileName, "w") as g:
    g.write("{}".format(neededRefrigerators))